### ISSUE TITLE
stm32duino port

### ISSUE THREAD

**Issue by ag88** [Chatanga/Girinoscope#3]
hi i'm thinking about adding a port for the girino arduino firmware to stm32duino 
http://wiki.stm32duino.com/index.php?title=Main_Page
https://github.com/rogerclarkmelbourne/Arduino_STM32
the intended board is maple mini or blue pill
https://www.leaflabs.com/maple
the boards are rather low costs
https://www.ebay.com/sch/i.html?_from=R40&_nkw=stm32f103c8t6&_sacat=0
the mcu is STM32F103C8 Arm Cortex M3
https://www.st.com/en/microcontrollers-microprocessors/stm32f103cb.html
72mhz arm cortex m3, 20k sram, 64k or 128k flash 
2 x 12-bit, 1 μs A/D converters (up to 16 channels) - in fast dual interleaved mode may reach 2 million samples per sec
would you be willing to handle pull requests for the java code?
the main things are:
- the base frequencies are 72mhz rather than 16mhz, the prescalers would be different (at least the base frequency is different) higher sampling rates
- i may want to support the 12 bits adc (i think currently 8 bits are used)
- that would mean bigger buffers possibly 16 bits x number of samples large, and Girino would need to handle different word sizes 
but i figure that some changes (possibly rather significant) may be needed on the java frontend
e.g. that we may need to abstract further and build girino 'factories' and need to use more parameters
e.g. base_frequency, frame_size, word_size, adc_bits, chip or device 
so that if the additional info is not present, fall back to the standard girino settings
at the moment i'm still prototyping and trying things out so nothing concrete as of now.

**Comment by ag88** [comment]
in addition reference voltage is different and only internal vref is available on the smaller low cost devices vrefint is 1.2v
stm32 is a family of mcus for now i'm only targetting the low cost pretty popular ones, the 'larger' ones has other vref such as external vref and some with even higher sampling speeds e.g. stm32f303 has 5msps adc x 4, in quad interleaved mode i've read posts showing that it is possible to push 20msps

**Comment by Chatanga** [comment]
Nice project.

Do not hesitate to push changes or even ask me to implement some of them. On the Java frontend side, a set of configuration files and an additional device type selection would be necessary. A dynamic switch to the proper device is even possible since the girino/stm32duino would probably identify itself at startup with a different message than "Girino ready".

For a start though, simply tweaking the constants in the Girino.java file should be enough (moving from 8 to 16 bits is the only thing which will require more changes).

**Comment by ag88** [comment]
thanks i'd try to make a 'girino compatible' one first and publish that on github
the incentive for supporting these 'blue pill' or 'maple mini' boards is
- low costs and widely available (just search stm32f103c8 or maple mini) on amazon, ebay, aliexpress and it turns out lots and lots of ads some as low as just usd $2 shipped
- it is a pretty fast cortex m3 arm processor 
- it has a good adc 1 msps x 2
- usb native (i.e. it is a generic usb device)
there are actually about 2-3 arduino cores (i.e. firmware cores for stm32)
1 https://github.com/rogerclarkmelbourne/Arduino_STM32 - this is a community core which originally starts form leaflabs
2 https://github.com/stm32duino/Arduino_Core_STM32 - this is actually from STM the company itself
3 https://github.com/danieleff/STM32GENERIC - this is another community based arduino core that is based around ST's HAL (hardware abstraction layer)
the cores (2) and (3) has possibly much more support for the stm32 family of processors as the HAL targeting the stm32 family processors is used, hence one would need to use part of the bulk HAL codes / library to compile it
the cores (1) is originally written by leaflabs and supports mainly/only the stm32f103. the community on www.stm32duino.com has extended that to support stm32f407. it is hence leaner and i'm using core (1) as i'm most familiar with it

**Comment by ag88** [comment]
i did an initial (draft) port 
https://github.com/ag88/GirinoSTM32F103duino
for this port it is 'Girino compatible' i.e. it works with the Girino frontend without changes.

a couple of things:
- i'd like to suggest that you could create a repository on github for the Girino arduino source as well (separate from the Java front end)
- i'd think it's good to create a protocol.md file to document the arduino protocol that communicates with the java frontend, it is pretty good actually, but i'd guess it is still in flux and would change, but that's ok

**Comment by Chatanga** [comment]
I’ve already thought about creating a repository for the Girino code, but discarded the idea considering its not my decision to take and that maybe its creator prefers to keep it at Instructables. That being said, it doesn’t matter since there is already such a repository you give an URL in your [README](https://github.com/ag88/GirinoSTM32F103duino/blob/master/README.md).

Regarding the protocol used by Girino, it does its job decently, but I wouldn't say that the protocol is "pretty good" considering its quirks and limitations, nor that it is still in flux. However, I didn’t try to change it to avoid any mandatory changes to the code. The optimizations I list are just that and remain optional.

> there is no comparator on stm32f103c{8,b}, signal triggering is done in software

What do you mean exactly by that? Is it implemented in firmware as some kind of pooling?

Lasty, I’ve spend some time thinking about how the code could be change to accomadate nicely with different Girino "profile". I’ve ended up with a question: how are you planning to transmit and potentialy store you 12 bits frames: in 16 bits aligned chunks or in a packed sequence of 12 bits? More precisely, is there a requirement for the latter?

**Comment by ag88** [comment]
hi thanks, i did not realise the Instructables for Girinoscope and you are really different authors :)
however, i'm nevertheless happy to work with you in part hopefully to extend the framework.
as stm32f103 mcu do not have a separate comparator, i made do by reading ADC values and doing the triggering in software. i.e. in my arduino codes i do a value comparison between adc reading and the threshold and did some logic to emulate a triggering as one would anticape. hence, in Griniscope java frontend, if you move the y-axis horizontal lever, it sets the threshold and i used that to create the 'triggering' action. it works pretty well when it tried it

**Comment by ag88** [comment]
note that the arduino core i'm using is the libmaple core https://github.com/rogerclarkmelbourne/Arduino_STM32
the boards i'm targeting is actually rather low cost and popular as they can be easily ordered on ebay, amazon etc. e.g.
https://www.ebay.com/sch/i.html?_nkw=maple+mini&_sacat=0
https://www.ebay.com/sch/i.html?_nkw=stm32f103c8t6&_sacat=0

**Comment by ag88** [comment]
note that the arduino core i'm using is the libmaple core https://github.com/rogerclarkmelbourne/Arduino_STM32
the boards i'm targeting is actually rather low cost and popular as they can be easily ordered on ebay, amazon etc. e.g.
https://www.ebay.com/sch/i.html?_nkw=maple+mini&_sacat=0
https://www.ebay.com/sch/i.html?_nkw=stm32f103c8t6&_sacat=0

**Comment by ag88** [comment]
note that the arduino core i'm using is the libmaple core https://github.com/rogerclarkmelbourne/Arduino_STM32
the boards i'm targeting is actually rather low cost and popular as they can be easily ordered on ebay, amazon etc. e.g.
https://www.ebay.com/sch/i.html?_nkw=maple+mini&_sacat=0
https://www.ebay.com/sch/i.html?_nkw=stm32f103c8t6&_sacat=0

**Comment by ag88** [comment]
i've discussed some further details about my implementation on the following forum threads
http://www.stm32duino.com/viewtopic.php?f=19&p=53911
https://mcu.selfip.com/viewtopic.php?f=10&t=26

**Comment by ag88** [comment]
one of the 'problems' as i discover is that as i pretty much do some kind of polling, actually not polling, i hooked the interrupt handler for the ADC end-of-conversion interrupt and another timer interrupt.
the end-of-conversion interrupt is where i did my 'magic' (i.e. i read the ADC value do that 'triggering logic', do voltage level conversions etc and puts that in a queue so that the front end can retrieve it).
and in addition, the voltage scales are different. hence, 
1) the voltages doesn't 'look correct'
on stm32 the adc outputs ranges from 0 - 4096 for 0 to 3.3v respectively (only 8 bits used to maintain compatibility). while the arduino implementation on atmega 328p has a range from 0 - 255 for -2.5v to 2.5v. hence, the voltage ranges would look 'incorrect', i.e. -2.5v on the graph is really 0v on stm32f103 and 2.5v is 3.3v on the stm32f103
2) i'm hit by a max sample rate of 200 k samples per sec due to my rather naive implementation. it is easier to code and understand, but at the max samplerates of 304 khz on Girinoscope, it simply looks incorrect as my implementation couldn't catch up.

**Comment by ag88** [comment]
nevertheless it is still a working and 'Girino compatible' implementation. 
for now i'd like to suggest we'd close this issue and made do with the 'Girino compatible' implementation as it would likely 'just works' with no changes to the frontend. i'd likely simply leave it at this state till i've perhaps more time to work on it.
to make better use of the stm32 platform, various changes on the java frontend are likely necessary but i've not thought too much in detail about it. e.g. about the voltage ranges mentioned prior and if we'd support the 12 bits adc rather than 8 bits. in addition, the 'automatic detection' of the implementation and parameters (e.g. voltage ranges and ADC number ranges, time scales and sample rates etc). but those would really take much more efforts. and also to leave room for 'future' additional implementations. the good thing about java is its object oriented/modular nature. hence we'd be able to accomodate 'different girinos' with interfaces and classes

**Comment by ag88** [comment]
nevertheless it is still a working and 'Girino compatible' implementation.
for now i'd like to suggest we'd close this issue and made do with the 'Girino compatible' implementation as it would likely 'just works' with no changes to the frontend. i'd likely simply leave it at this state till i've perhaps more time to work on it.
to make better use of the stm32 platform, various changes on the java frontend are likely necessary but i've not thought too much in detail about it. e.g. about the voltage ranges mentioned prior and if we'd support the 12 bits adc rather than 8 bits. in addition, the 'automatic detection' of the implementation (e.g. the 'Girino ready', we'd likely need different tags e.g. 'Girino stm32f103mm ready'. note initially when i tested it Girino front end did not manage to read that 'Girino ready', a more robust implementation would be for the front end to issue a command (ask) and the 'girino' firmware respond to the command) and parameters (e.g. voltage ranges and ADC number ranges, time scales and sample rates, identification (i.e. the name e.g. stm32f103mm and a version etc). but those would really take much more efforts. and also to leave room for 'future' additional implementations. the good thing about java is its object oriented/modular nature. hence we'd be able to accomodate 'different girinos' with interfaces and classes

**Comment by ag88** [comment]
nevertheless it is still a working and 'Girino compatible' implementation.
for now i'd like to suggest we'd close this issue and made do with the 'Girino compatible' implementation as it would likely 'just works' with no changes to the frontend. i'd likely simply leave it at this state till i've perhaps more time to work on it.
to make better use of the stm32 platform, various changes on the java frontend are likely necessary but i've not thought too much in detail about it. e.g. about the voltage ranges mentioned prior and if we'd support the 12 bits adc rather than 8 bits. in addition, the 'automatic detection' of the implementation (e.g. the 'Girino ready', we'd likely need different tags e.g. 'Girino stm32f103mm ready'. note initially when i tested it Girino front end did not manage to read that 'Girino ready', a more robust implementation would be in addition to 'Girino ready' for the front end to issue a command (ask) and the 'girino' firmware respond to the command if a timeout occurs) and parameters (e.g. voltage ranges and ADC number ranges, time scales and sample rates, identification (i.e. the name e.g. stm32f103mm and a version etc). but those would really take much more efforts. and also to leave room for 'future' additional implementations. the good thing about java is its object oriented/modular nature. hence we'd be able to accomodate 'different girinos' with interfaces and classes

**Comment by ag88** [comment]
oh in addition, think about multi channel, multi graphs, i think it is a rather popular request as i've a 'user' ask about it. i'd think a main idea is to abstract the 'girino' and the 'frontend' parts into java classes and interfaces and in addition on the girino parts to gradually form a (serial) 'protocol' as like the current girino protocol. most likely in the form of commands (from frontend) and responses and actions (from the girino firmware).

**Comment by ag88** [comment]
oh in addition, think about multi channel, multi graphs, i think it is a rather popular request as i've a 'user' ask about it. i'd think a main idea is to abstract the 'girino' and the 'frontend' parts into java classes and interfaces and in addition on the girino parts to gradually form a (serial) 'protocol' as like the current girino protocol. most likely in the form of commands (from frontend) and responses and actions (from the girino firmware).

**Comment by ag88** [comment]
further works from this point is likely to be a 'fork' and i may call it girino stm32f103mm port v2 or something, the existing port would be renamed the 'compatible' release. i would likely keep both set of codes in the same repository in different folders

the fork is likely necessary as those changes would likely make it no longer 'girino compatible' (e.g. possible use of full 12 bits, communicate parameters of voltage ranges to number range mapping, sample rates, features etc)
i'd think for display scales (voltage & time) that is a gui prerogative, but user should probably be allowed to vary the voltage scaling display and time scales. i.e. The notion is of zooming, i.e. a user may want to zoom in to parts of a waveform etc.

we can use a new issue ticket as a 'thread' for discussions or we can discuss this say in a forum thread like
https://mcu.selfip.com/viewtopic.php?f=10&t=26

**Comment by ag88** [comment]
some good news, i rewrote much of the existing codes and did an accelerated version of girino stm32duino. it can now handle the max sample rates of 307.7 k samples per sec in girinoscope currently without issues.
the source is in the Girino_stm32f103mm_v2_compat folder. this is still a 'compatible' release
it is currently still draft, things may be broken
some details in my commit message
https://github.com/ag88/GirinoSTM32F103duino/commit/d76617f5d773f5e37c7a0a4f2a0392f5232c4f13

**Comment by ag88** [comment]
i've released a version 2 port, this is still a 'compatible' release
updated readme.md on the repository as well
https://github.com/ag88/GirinoSTM32F103duino

Girino_stm32f103mm_v2_compat is the current release
accel girino compat v2 using DMA and AWD, beta
This is an accelerated girino compat release, beta.
Near complete re-write. Now girino stm32duino v2 is a significantly accelerated using stm32 hardware DMA and analog watchdog

Fixed issue related to the setting of wait duration

It now has no issues displaying signals the highest sample rates currently of Girinoscope 307.7 k samples per sec

Did further enhancement and achieved 500 k samples per sec - limit (this 'feature' is not obserable on the java frontend in compatable mode).
internally it is already doing a full 12 bits ADC buffers 8 bits interfaced to Girinoscope for compatibility

This is a beta release

**Comment by Chatanga** [comment]
I’ve pushed a new unstable version of the Girinoscope which supports multiple device profiles. You should now be able to introduce (limited) incompatible changes with the original Girino firmware. I’ve already add an entry for the STM32duino, but tweaking the [code](https://github.com/Chatanga/Girinoscope/blob/master/src/main/java/org/hihan/girinoscope/comm/Device.java) is probably necessary. I didn’t test it anyway.

Regarding your comment on the display scale (the Y axis at least), haven’t you noticed that the current version already have a "Display / Change signal interpretation menu" option to change it? The [-2.5, 2.5] has always been arbitrary, even for the Girino. The good thing with the new version, is that it save your settings and you just need to set your display range once.

You also mention the fact that the frontend is not always waiting for the 'Girino ready' properly. The expected behavior is for the Girino to be reset when the USB / serial connection is etablished, leaving it with a WAIT_DURATION delay to initialize before the frontend complain about not getting any readiness message. Depending on you USB / serial adapter, this reset could not happen. That being said, having some kind of "ask" command, as you suggest, would be more robust.

Lastly, I’m not sure that implementing multiple channels and graphs is in line with the original intent behing the Girino: squeeze the fastest sampling rate from the original Arduino. Obviously, if you use a more powerfull device, you will be able to finally start to have something more akind to a real oscilloscope (as opposed to a basic Data Logger in disguise).

**Comment by ag88** [comment]
hi Chatanga thanks! i'd check that out. at the moment i've been doing 'compatible' releases to figure out what are the limitations. As of current, the most recent release which is still a 'compatible' release maps pretty much the key functionalities of Girinoscope. it is now able to display signals at  307.7 ksps sample rates (the earlier release hit limits) and now both the levers for threshold vertical axis and horizontal axis the 'wait duration' which i mean works. As you have mentioned in your readme.md there is a bug in the "wait duration' parameter using uint8_t which limits that value to 255 which is 'incorrect'. i've used uint16_t which allows the full width of 1280 to be specified as needed.
in my implementation, in the current design the 's' command initiates the conversion and reverts with a full frame (1280 items) of data and stop. this is so that conversions is driven purely by the 's' command and it stops when the frame is filled, leaving an allowance for the 'trailing' parts controlled by wait duration.

i realised that one of the trouble is that it is not possible for you to target an unknown feature set.
in particular, i do not specifically have a 'firmware' for you to literally test and to figure out how that works. Unfortunately, neither do i know how my eventual feature sets is going to look like as i'm still figuring out how to do them. 

Because of this iterative improvement nature of things, i'm doing so by achieving the current status quo, e.g. it 'works' without changing the java front end codes.  as of the current release, this is achieved.

stm32 F1 series (cortex-m3) and stm32 F4 series (cortex m4) are resourceful and 'powerful' mcus.
e.g. for the f103 the adc is capable of monitoring 16 channels (round robin polling) at a net max of 1 msps based on specs and there is actually 2 ADC on the stm32f103c8 (https://www.st.com/en/microcontrollers-microprocessors/stm32f103cb.html). and a stm32f407ve (https://www.st.com/en/microcontrollers-microprocessors/stm32f407ve.html) has 3 ADC (multi channel sampling) with 2.4msps per ADC based on spec. hence if all 3 ADCs are used the combined throughput is some 7.2 msps, which makes multi channel display looks pretty feasible.

in reality, when i start working with this project, i found that there are many intervening factors. for instance my initial naive implementation only achieved 200 k samples per sec max on a stm32f103.
while the current hardware accelerated release i managed to push 500 k samples per sec during my tests. And in addition this time round there is a difference from the atmega328p, that 500 k samples per seconds is driven by elaborate hardware timers as opposed to simply using a pre-scaler divider.
hence, stm32f103 can do 0 - 500 k samples per seconds in a continuous specification i.e. any number you pick as sample rates within that range. a divide by 2 pre-scaler would normally leave a lot of gaps between the sample rates.

however, as like discussed earlier, it isn't possible for you to target an 'unknown' feature set. And unfortunately i'm also trying to figure out what that 'unknown' feature set looks like (i.e. i'm exploring and developing it). i'd review the java codes and design and to suggest the 'features' as i go along and i'd keep in mind that it shouldn't 'break' the framework in ways that it causes large changes to be necessary. (i.e. if that isn't possible then it'd be better i start a different and new java front end to keep things 'simple'). hope this helps clarify things

**Comment by ag88** [comment]
in addition, you could also fork my repository on your github, and i welcome pull requests and suggestions from you as well :)

**Comment by ag88** [comment]
hi Chatanga, 
it certainly looks like quite a lot of work ! thanks :)
i'm trying out the new experimental 'STM32duino (experimental)'  device. 
it seemed the voltage reference parameter is marked 'fake', so initially i hit a bit of stall, and it turns out the frontend is waiting for the missing (voltage reference) parameter as it is trying to read that :p lol (i've added a 'voltage reference: 0' on the firmware for now to get around it, only on my test copy)

oh well i saw the base frequency 72mhz, that gives a top sample rates of 1.3msps, oh well i think it may be somewhat steep to live up to that for now :p lol

i see the new frame format class, there may be some challenges on the mcu end.
the adc is 12 bits, but the buffer may be 16 bits based. 

i'm thinking more in terms of having an abstract Grino class and possibly Device class.

this would be so that you may have a DeviceFactory or GirinoFactory class.
then Girino inherits / extends the AbstractGirino class.
in the same way my 'Girino' inherits / extends the AbstractGirino class and override the hooks.
in that sense it would be int[] AbstractGirino.acquireData(), i.e. the int would contain the adc raw data, 1 sample per int or you could literally ask for an array of double[] where i give you the volts directly.

one of those things is that for PrescalerInfo there would be 2 types on stm32.
- fixed sample rates (e.g. based on prescaler selection) - this would be one type, 
  currently the girino commands are like pN where N is the desired prescaler
- range sample rates with an allowable range min sample rates to max sample rates, the user is allowed to choose any integer sample rates within the range, 
this may look like Pnnnnn where nnnnn is the direct desired sample rates. the units could be hundreds of sample per sec. e.g. P5 is 500 samples per sec and P50 is 5 k samples per sec etc.
i may change the protocol to one with an end of line linefeed or carriage return, not decided yet. but possibly that may allow more flexible command and parameter specs going forward.

i'm currently still struggling to get myself familiar with the codes. Hence, don't be too bothered by this i may take some time to get up to speed.

**Comment by ag88** [comment]
note that on stm32 'fixed' prescalers are not prescalers, they are categories of fixed sample rates.
e.g. like 1 may be for 600 sample per sec, and 2 for 857 samples per sec etc. these are achieved by setting the adc on 'continuous' mode (i.e. covert a new sample once the current is done)

i may add them later, but for now the achievable range is 0 - 500 k samples range sample rates where the user can choose a value. this is achieved on the mcu by using elaborate timers so that any of the sample rates within the achievable range can be selected

**Comment by ag88** [comment]
i'm thinking that some of the classes e.g. like Parameter and PrescalerInfo could be taken out and be their own classes rather than nested within Girino

**Comment by ag88** [comment]
Parameter could stay in Girino as it is an enum, apparently the isn't an 'easy' way to extend an enum other than using interfaces. i.e. if i create my Girinoxxx class, i'd have to define the Parameter enum (perhaps with new and different parameters) as well may likely be different.

**Comment by ag88** [comment]
Girino class could have a method that returns a list/set/array of PrescalerInfo so that the menu can be constructed, this isn't just Girino specific but device specific. i.e. the list of support fixed and range sample rates. 
e.g. 
    public static class PrescalerInfo {
     enum Type {
          FIXED,
          RANGE
     } type;
     int range_max;
     int rage_min;
     public final int value;
     public final double frequency;
     public final double timeframe;
     public final String description;
     public final boolean tooFast;
     public final boolean reallyTooFast;    
... etc

And PrescalerInfo could be use as an object to set the sample rates e.g. Girino.setsamplerate(PrescalerInfo pi), and Grino use that structure to determine the appropriate command to set the sample rate.

**Comment by ag88** [comment]
preferably the abstract Girino class provides additional methods like
- setwaitduration(int duration) where duration is number of samples to capture from trigger < frame_size
- settriggertype(TriggerType t) where TriggerType is an enum {
Toggle, RisingEdge, FallingEdge, None} for none, when my board receives the 's' (start converting) command it immediately start converting / capturing, i.e. it doesn't trigger but capture a full frame of data on 's' command
-setthreshold(int threshold) where threshold is the threshold level to trigger, this could be in terms of the adc levels or voltages (e.g. using double) but a direct spec using int is probably prefered as that's what is on the adc
... etc
--
by using an abstract Girino class interface it could help to provide a common API style interface between the Gui parts of Girinoscope and the Girino 'driver' parts

future additions could include abstract methods like 'channels' 
e.g. Girino.setchannel(int channel, boolean onoff)
and for those cases Girino.acquiredata() could return a 2 dimensional array or lists of lists.
e.g. returns ArrayList<ArrayList> channels, and data = ArrayList<Integer> channels.get(channel)

**Comment by ag88** [comment]
and a GirinoFactory class would mainly be a class creating the AbstractGirinos or Girino objects, i.e. GirinoFactory class would likely register all the different Girinos (and in future you could even implement plugins) and it may possibly do the 'smarts' e.g. read the different 'Girino xxx ready' and query the different Girino classes to identify it and return the relevant Girino object.

To implement my Girino xxx, i'd inherit from AbstractGirino class and override all the necessary class methods and i'd possibly edit GirinoFactory so that it would register my Girino xxx in its 'Girino lists'

**Comment by ag88** [comment]
what seemed missing is also a single capture mode where the firmware does the triggering and capture of one frame of data. but the front end stops after a single 's' (start conversion) command. that would allow impluse situations to be observed, e.g. charging a capacitor with a on switch etc

**Comment by ag88** [comment]
i realise that a key part of the 'magic' are the parameters and that each parameter is literally some kind of command pattern https://en.wikipedia.org/wiki/Command_pattern
this is apparently necessary as when the user change a parameter, an acquisition may be on-going, hence the commands need to be queued  

my AbstractGirino mentioned above in this case is more of an abstraction of a device.
i'm thinking that the parameters list can be maintained by an instance of AbstractGirino.
if that is the case each parameter would become an instance variable at least in AbstractGirino.

in that case, a paramter isn't simply an enum in the logical sense.
For instance, the parameter TRIGGER_EVENT command 'e' would be related to TriggerEventMode and its various modes. In that sense, it would seem more appropriate to have an AbstractParameter as a parent class, so that TriggerEventParameter inherits from it and TriggerEventMode is part of the enums of ParameterTriggerEvent. But this thinking may lead of quite significant code re-organisation.
and so that the 'parameter list' is actually a list of AbstractParameters.

and gui codes could literally call, Girino,getTriggerModes() to say get a list of trigger modes to setup the menu and when the user selects it call Girino.setTriggerMode(ParameterTriggerEvent.TriggerEventMode mode)
while the Girino object instance maintains the parameters list
but in the same way the codes would no longer look the same as current if we proceed with the 're-org / re-factoring'

I'm currently not changing any codes and i'd not propose any PR unless you are comfortable with the re-org. In fact, i'd think it would be better that i leave the re-org to you so that you could decide how the design and re-org should be.

unfortunately an 'API' like interface (e.g.. an AbstractGirino class and methods / instance variables and behaviour) would define the notion of what is a 'Girino') this becomes a design decision in part and it may be more appropriate for you to decide on them.

**Comment by ag88** [comment]
oh on the side there is also the java service provider interface
https://docs.oracle.com/javase/tutorial/ext/basics/spi.html
which you could do things like Iterator<Girino> girinos = loader.iterator();
to visit all the Girinos at initialisation, this could go into that GirinoFactory class 
so that its constructor can setup a 'Girino list' at initialization, e.g. it could call a method like Girino.getIdentity() and the AbstractGrino needs to respond appropriately. That is just the class, not interfacing to serial yet.

just 2 cents

**Comment by ag88** [comment]
hi don't change anything just yet, i'm figuring out some of the limitations of the stm32f103 adc i could possibly create a setup that fit/conform to the current framework but yet provide better sample rates with the new custom stm32f103mm device. if i manage to do that i'd propose some changes within the current framework. i.e. without the abstract girino class

**Comment by ag88** [comment]
if you are going ahead with the abstract girino, don't hestitate to do so just post a note so i'd change directions again, my thoughts are an abstract girino could be a 'longer term' solution :)

**Comment by Chatanga** [comment]
I’ve pushed a minor change in the repository which removes the FAKE type for a parameter (it was mostly useless). I’ve kept the prescaler terminology around but, for the STM32duino, I agree it is misleading. That being said, nothing prevents you from providing any kind of values and descriptions. The only restriction is that it is only a limited discrete set of values, not a continuous range of values.

In addition, I’ve also designed a draft for [a more generic API](https://gist.github.com/Chatanga/0be36d0011bcc619499fcddf357bee9d). Forget the Device and FrameFormat classes and think about Girino class as an implementation, among others, of this interface (and maybe an AbstractGirino to reuse code). This new API obviously requires more works on the UI side and could take time to implement. In fact, since I’m also interrested in playing with JavaFX (and JDK 11), I could also possibly rework a lot more things ;) Better start with a well-though API in this case.

**Comment by ag88** [comment]
thanks!
the new generic api looks quite good but i'm thinking you could literally use an abstract class rather than an interface. that could allow some varables to be drawn up in the abstract class, e.g. we could have some of the parameters represented as variables. e.g. Parameter baseSampleRate, TriggerEvent (toggle, rising, falling), waitDuration (number of samples), Threshold, VoltageRange (the y axis), FrameSize (in number of samples), Channels [] (possibly we could use a list or something)
in terms of a model i took a look at this picture of an oscilloscope, i'm hoping this could help simplify things somewhat. i'm not too sure if anything else would group into that Channels []
https://commons.wikimedia.org/wiki/File:Oscilloscope_Front_Panel.svg
on the stm32 hardware/firmware front, what we are really seeing is the ADC
and for VoltageRange i'd make it represent what the ADC sees e..g 0 (0v) - 4096 (3.3v 12 bits),

on the design front, i go with the notion that one frame is a *fixed number of samples* (1280). so a single sweep with the 's' (start conversion) command would return that number of samples. as each sweep with the 's' command is typically short it seem unlikely that we'd need to return a partial frame. don't think classic girino do that either.

i think a simplifying assumption we are making here is that on the horizontal axis it would be a common base sample rate. e.g. if we have 2 channels, they share the same prescaler or base_sample_rate. so we could think of a 2 dimensional array with each having 1280 samples.
(the number of samples possible may be restricted by available ram when both channels are selected, but i've not thought too deeply about it, but that at the point when that is set, the FrameSize would be less than 1280 samples)

i'm confronting various challenges on the mcu front as well while i'm working on this
the lower sample rates 0 - 500 k samples per sec, stm32f103 could deliver a 'range' sample rates.
but above that, things become discrete, e.g. i may be only able to offer discrete rates like 600 ksps, 857 ksps, and maybe 1.714 msps, still experimenting. in that sense, to setup the sample rates (prescaler) menu, the call to Girino.getAvailableSampleRates. i may need to return a list of rates like
1.714 msps, fixed
857 ksps, fixed
600 ksps, fixed
0 - 500 ksps, range

if you prefer to set the sample rates say using Girino.setBaseSamplerates(int samplerates), what i could do could be to find the (nearest) value fitting the selections offered. in that sense setBaseSamplerates() could return the actual sample rates set and/or update the base_sample_rates instance variable similarly

just some thoughts,
cheers :)

**Comment by ag88** [comment]
oh that 1.714 msps is a 'special' mode which i'm still exploring
it works by means of using 2 ADC in 'fast interleave' mode.
i.e. if you need 2 channels, it may become 857 ksps (fixed) instead in addition to the limitation on available mcu sram memory. i.e. when settings are changed, dependent parameters such as sample rates and frame_size may be changed / lowered, i've not thought too deeply about this either

it is inevitable that limits are reached pretty quickly as that is a rather 'small' mcu with 20k sram with its built-in ADC which while rather versatile certainly has its hardware design limits and capabilities limits

**Comment by Chatanga** [comment]
**About the abstract class**

Let’s clarify things up: the Parameter and ParameterValue interfaces are not just there because I’ve also introduced a Driver interface. There won’t disappear and become "some varables to be drawn up in the abstract class" in case an abstract class is introduced. They are here to address the fact that implementations could introduce new parameters with any kind of values. To be honest, the sole parameter requiring this today is the prescaler/samplingRate (but it wont be enough with you continuous-then-discrete you are now describing...). I’ve decided to generalize it to all parameters "just in case" (who know? Maybe the Globoscope3000 will introduce a continuous trigger mode based on the increase speed of the signal). It could well be over-engineered and it would be a valid point to say so, but it has nothing to do with abstractness.

_(You could skip this paragraph and jump to the next.)_

In OOP, the best usage of abstract classes is not to use them. That’s a bit provocative, but an API shouldn’t rely on abstract classes, because abstract class have a tendency to break encapsulation. That a given implementation (in the broad sense) uses abstract classes internally to share code is ok, exposing such class to be extended by other people in other (unknown) contexts is seldom a good idea. Of course, you could design an API with extensible classes (eg. Swing components), but it’s more about customization than code sharing. The rule here is that if you need to know how your abstract class works internally (and it is your case), you are not encapsulating things properly. That is the theory, the Girinoscope remains a small application, little more than a toy, and an AbstractGirino could be introduce to simplify things (but a kind of PropertyFactory is probably more useful I think). However, the interface remains valuable here. In fact, there is always an interface even if not reified into a "Java interface".

**About the channels**

I’m still seeing the Girinoscope as a datalogger, not a proper oscilloscope (and as long as it wont be able to send continuous data in real time, it wont be one). So, when you said 2 channels, I didn’t think about the classic X/Y view. In this configuration, I agree that having more than 2 channels doesn’t make sense (are 3D oscilloscope a thing today ?) or only in a X/Y1..YN configuration. That being said, it doesn’t really change the API. The time information becomes useless, but it is more a UI issue to display frames in X/Y or not (or even in 3D, could be cool).

**One last thing**

Why do you "go with the notion that one frame is a fixed number of samples (1280)"? The Girino uses this fixed size, your implementation could do the same for comparable reasons, but it doesn’t matter at all on the UI (and API) side. It is no like the API is enforcing a writable frameSize parameter (but an implementation could provide it as a generic parameter if needed).

**Comment by ag88** [comment]
thanks, neither could stm32f103 send continuous data in real time, hence, it would also be 'just another datalogger' and pretty much another toy. the way i'm seeing things is that using the 's' (start conversion) command the device collects a fixed number of samples (a frame) (what may possibly be changed are the sample rates) and sends it to the host app, in this sense it emulates the sweep on an oscilloscope. however, the next sweep while close could have a gap (and serial do have speed limits), and it relies on the trigger which would capture the next arbitrary edge and send the next set (frame) of data. so a repetitive signal would seem rather stationary that way.

abstract class
i think your point is pretty much valid and good, i go with the notion of those variables as i thought it would give more form of a model hence i actually hope that could help simplify things. but i'm just as much a novice as any where it comes to this. i'd like to leave this with you 

channels
i think it started with my overzealous mention about channels :p lol
your notion of girinoscope as a datalogger is a good one. i initially started with a picture of a 2 channel oscilloscope sharing a same time base as implementation wise that is about what is possible on a single stm32 small mcu. In addition i think that this may be a useful simplification.
however, if we consider channels as arbitrary data logs and is not necessarily constrained by a same time base, it opens up new and possibly useful scenarios, e.g. you could literally have 2 arduino devices feeding data (at possibly different sample rates), and that would still constitute 2 channels of data. but now channels isn't simply a 2 x frame size set of numbers that we'd plot on a same scale. and to an extent now channels would take on more attributes e.g. sample rates. in the same way, i'd leave this to you

**Comment by ag88** [comment]
closing this as this would be a 'feature request' and not a 'bug'
